pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+type engine struct {
+	uuid [16]byte
+	conn pgxconn.PgxConn
+
+	stop chan struct{}
+
+	metrics sync.Map
+}
+
+type Telemetry interface {
+	Start()
+	// RegisterMetric registers a Prometheus Gauge or Counter metric for telemetry visitor.
+	// It must be called after creating the telemetry engine.
+	RegisterMetric(columnName string, gaugeOrCounterMetric prometheus.Metric) error
+	Stop()
+	// Sync syncs the registered metric values with the information table. This is used only
+	// in end-to-end tests.
+	Sync() error
+}
+
+func NewTelemetryEngine(conn pgxconn.PgxConn, uuid [16]byte) (Telemetry, error) {
+	isTelemetryOff, err := isTelemetryOff(conn)
+	if err != nil {
+		log.Debug("msg", "unable to get TimescaleDB telemetry configuration. Maybe TimescaleDB is not installed", "err", err.Error())
+	}
+	if !isTelemetryOff {
+		// Warn the users about telemetry collection only if telemetry collection is enabled.
+		log.Warn("msg", "Promscale collects anonymous usage telemetry data to help the Promscale team better understand and assist users. "+
+			"This can be disabled via the process described at https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry/#disabling-telemetry")
+	}
+	engine := &engine{
+		conn: conn,
+		uuid: uuid,
+	}
+	if err := engine.writeMetadata(); err != nil {
+		return nil, fmt.Errorf("writing metadata: %w", err)
+	}
+	return engine, nil
+}
+
+func isTelemetryOff(conn pgxconn.PgxConn) (bool, error) {
+	var state string
+	err := conn.QueryRow(context.Background(), "SHOW timescaledb.telemetry_level").Scan(&state)
+	if err != nil {
+		// Maybe TimescaleDB is not installed, in that case, we cannot send telemetry, hence don't log it.
+		return true, fmt.Errorf("fetching timescaledb telemetry setting: %w", err)
+	}
+	if state == "off" {
+		return true, nil
+	}
+	return false, nil
+}
+
+// writeMetadata writes Promscale and Tobs metadata.
+func (t *engine) writeMetadata() error {
+	promscale := promscaleMetadata()
+	if err := writeToTimescaleMetadataTable(t.conn, promscale); err != nil {
+		return fmt.Errorf("writing metadata for promscale: %w", err)
+	}
+
+	tobs := tobsMetadata()
+	if len(tobs) > 0 {
+		if err := writeToTimescaleMetadataTable(t.conn, tobs); err != nil {
+			return fmt.Errorf("writing metadata for tobs: %w", err)
+		}
+	}
+	return nil
+}
+
+const (
+	metadataUpdateWithExtension = "SELECT update_tsprom_metadata($1, $2, $3)"
+	metadataUpdateNoExtension   = "INSERT INTO _timescaledb_catalog.metadata(key, value, include_in_telemetry) VALUES ('promscale_' || $1, $2, $3) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value, include_in_telemetry = EXCLUDED.include_in_telemetry"
+)
+
+// writeToTimescaleMetadataTable syncs the metadata/stats with telemetry metadata table and returns the last error if any.
+func writeToTimescaleMetadataTable(conn pgxconn.PgxConn, m Metadata) error {
+	update := func(queryFormat string) error {
```

Comment:
What's the reason for having this as an inline function? Don't see much of a reason and it hurts readability by increasing nesting in this function.

Reply:
Because there isn't any good reason to write a new function block just to reduce redundancy. Sorry, but this is a very simple function, which just updates the table. Otherwise, we will have to write a func that accepts `conn, m, queryFormat` which I don't think we wanna do.